rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -0,0 +1,147 @@
+        logger.error(f"{prefix} Output [{output_name}] has no subsequent node.")
+        return None
+
+    next_node_id = node.output_nodes[output_name]
+    next_node = graph.get_node(next_node_id)
+    if not next_node:
+        logger.error(f"{prefix} Error, next node {next_node_id} not found.")
+        return None
+
+    next_node_input = graph.get_node_input(next_node, run_id)
+    next_node_block = next_node.block
+
+    if not set(next_node.input_nodes).issubset(next_node_input):
+        logger.warning(
+            f"{prefix} Skipped {next_node_id}-{next_node_block.name}, "
+            f"missing: {set(next_node.input_nodes) - set(next_node_input)}"
+        )
+        return None
+
+    if error := next_node_block.input_schema.validate_data(next_node_input):
+        logger.warning(
+            f"{prefix} Skipped {next_node_id}-{next_node_block.name}, {error}"
+        )
+        return None
+
+    logger.warning(f"{prefix} Enqueue next node {next_node_id}-{next_node_block.name}")
+    return Execution(run_id=run_id, node_id=next_node_id, data=next_node_input)
+
+
+def execute_node_sync(data: Execution) -> Optional[Execution | None]:
+    """
+    A synchronous version of `execute_node`, to be used in the ProcessPoolExecutor.
+    """
+    prefix = get_log_prefix(data.run_id, data.id)
+    try:
+        logger.warning(f"{prefix} Start execution")
+        return execute_node(data)
+    except Exception as e:
+        logger.error(f"{prefix} Error: {e}")
+
+
+class ExecutionManager(AppService):
+    def __init__(self, pool_size: int):
+        self.pool_size = pool_size
+        self.queue = ExecutionQueue()
+
+    def run(self):
+        db.connect()
+        block.initialize_blocks()
+
+        def on_complete_execution(f: asyncio.Future[Execution | None]):
+            exception = f.exception()
+            if exception:
+                logger.exception("Error during execution!! %s", exception)
+                return exception
+
+            execution = f.result()
+            if execution:
+                return self.__add_execution(execution)
+
+            return None
+
+        with ProcessPoolExecutor(
+            max_workers=self.pool_size,
+            initializer=db.connect,
+        ) as executor:
+            logger.warning(f"Execution manager started with {self.pool_size} workers.")
+            while True:
+                future = executor.submit(execute_node_sync, self.queue.get())
+                future.add_done_callback(on_complete_execution)  # type: ignore
+
+    @expose
+    def add_execution(self, run_id: str, node_id: str, data: dict) -> dict:
+        try:
+            execution = Execution(run_id=run_id, node_id=node_id, data=data)
+            return self.__add_execution(execution).dict()
+        except Exception as e:
+            raise Exception("Error adding execution ", e)
+
+    def __add_execution(self, execution: Execution) -> Execution:
+        add_execution(execution)
```

Comment:
Maybe rename one of these `add_execution` functions. 3 versions of `add_execution` seems a little its going to lead to a bit of confusion.

Reply:
Good point, I renamed the add_execution on the model layer to enqueue_execution